app/controllers/api_controller.rb

```diff
@@ -28,6 +30,20 @@ def user_for_paper_trail
 
 private
 
+  # NB: for new controllers
+  def validate_required_idempotency_key
+    idempotence_validator.call(request.headers['IDEMPOTENCY_KEY'], required: true)
```

Comment:
Im slightly confused by the validator's function: it will set a new `idempotency_key` if it doesn't exist or throw an error otherwise. But if we were to send this request twice with the same key, I thought we should not see an error but rather return the existing key (and later on not perform the same action because it exists)?

Reply:
so, the function of the validator is: accept the request if it has not already been received before.

the test is: is the key in redis? 
if not, add the key and allow the request. 
If the key is in redis, reject the request with a conflict error.

Reply:
Could we return success but not perform any of the side effects? I thought that was the point of being idempotent?

Reply:
@willfish no, the expected behaviour (as specified in supplier API specifications, swagger and other docs) is to return an HTTP 409 Conflict error.
Any change to this behaviour (to return success rather than error) would require discussion with suppliers as they will be coding with the current behaviour in mind.